lib/formatter.js

```diff
@@ -0,0 +1,72 @@
+'use strict';
+
+const generateFunction = require('generate-function');
+
+function getFormatter(metric, defaultLabels) {
+	const defaultLabelNames = Object.keys(defaultLabels);
+	const name = metric.name;
+	const help = escapeString(metric.help);
+	const type = metric.type;
+	const labelsCode = getLabelsCode(metric, defaultLabelNames, defaultLabels);
+
+	const gen = generateFunction();
+	gen(`
+  function format(item, escapeLabelValue, getValueAsString) {
+    const info = '# HELP ${name} ${help}\\n# TYPE ${name} ${type}\\n';
+    let values = '';
+    for (const val of item.values || []) {
+      val.labels = val.labels || {};
+      let metricName = val.metricName || '${name}';
+      const labels = \`${labelsCode}\`;
+      const hasLabels = labels.length > 0;
+      metricName += \`\${hasLabels ? '{' : ''}\${labels}\${hasLabels ? '}' : ''}\`;
+      let line = \`\${metricName} \${getValueAsString(val.value)}\`;
+      values += \`\${line}\\n\`;
+    }
+    return info + values;
+  }`);
+	return gen.toFunction();
+}
+
+function getLabelsCode(metric, defaultLabelNames, defaultLabels) {
+	let labelString = '';
+	const labelNames = getLabelNames(metric, defaultLabelNames);
+	for (let index = 0; index < labelNames.length; index++) {
```

Comment:
⭐ It's valid to only set values for a subset of labels when a metric is updated. With this test:

```js
it('should allow subsetse of labels', () => {
	instance = new Gauge({
		name: 'name_2',
		help: 'help',
		labelNames: ['code', 'success'],
	});
	instance.inc({ code: '200' }, 10);
	const str = globalRegistry.getSingleMetricAsString('name_2');
	console.log(str);
});
```

in this PR, you get 
```
# HELP name_2 help
# TYPE name_2 gauge
name_2{code="200",success="undefined"} 10
```

whereas in master, you get
```
# HELP name_2 help
# TYPE name_2 gauge
name_2{code="200"} 10
```

Reply:
should add this as a test, no?

Reply:
Yes, of course. I’ll fix it

Reply:
The above test case passes, but now if the first label is undefined, there's a stray leading comma:

```
it.only('should allow combinations of labels', () => {
	instance = new Gauge({
		name: 'name_2',
		help: 'help',
		labelNames: ['code', 'success', 'ok'],
	});
	instance.inc({ code: '200' }, 10);
	instance.inc({ code: '200', ok: 'yes' }, 10);
	instance.inc({ ok: 'yes', success: 'false' }, 10);
	instance.inc({ success: 'false', ok: 'no' }, 10);
	const str = globalRegistry.getSingleMetricAsString('name_2');
	console.log(str);
});
```

prints

```
      # HELP name_2 help
      # TYPE name_2 gauge
      name_2{code="200"} 10
      name_2{code="200",ok="yes"} 10
      name_2{,success="false",ok="yes"} 10
      name_2{,success="false",ok="no"} 10
```

(notice last two lines)

Something like these should both be committed as test cases. Do you feel like adding them please?

---

Just checked, benchmarks still look significantly faster despite the new branch added in the last commits. 👍